File: xbrl.py
#!/usr/bin/env python3
"""
Look up ESEF/XBRL filings for a company on filings.xbrl.org and print
a formatted table.

Usage:
    python3 xbrl_filings_lookup.py ASML
    python3 xbrl_filings_lookup.py "Interactive Brokers"
    python3 xbrl_filings_lookup.py ASML --limit 10
"""

import argparse
import json
import sys
import urllib.parse
import urllib.request

API_URL = "https://filings.xbrl.org/api/filings"

HEADERS = {
    "User-Agent": "Mozilla/5.0 (X11; Linux x86_64; rv:150.0) Gecko/20100101 Firefox/150.0",
    "Accept": "application/json, text/javascript, */*; q=0.01",
    "Accept-Language": "en-GB,en;q=0.9",
    "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
    "X-Requested-With": "XMLHttpRequest",
    "Connection": "keep-alive",
    "Referer": "https://filings.xbrl.org/",
    "Sec-Fetch-Dest": "empty",
    "Sec-Fetch-Mode": "cors",
    "Sec-Fetch-Site": "same-origin",
}


def build_url(query: str, limit: int) -> str:
    # Reproduces the exact filter structure the filings.xbrl.org frontend
    # itself sends (captured via browser devtools), unmodified. Do not trim
    # the "or" clauses or drop the UAIFRS exclusion -- both were present in
    # the confirmed-working request.
    filt = [
        {
            "or": [
                {"name": "entity.name", "op": "ilike", "val": f"%{query}%"},
                {"name": "entity.identifier", "op": "eq", "val": query},
                {"name": "language.name", "op": "ilike", "val": f"{query}%"},
                {"name": "sha256", "op": "eq", "val": query},
            ]
        },
        {"name": "input_filing.filing_source.program", "op": "ne", "val": "UAIFRS"},
    ]
    params = {
        "include": "entity,language",
        "filter": json.dumps(filt),
        "sort": "-date_added",
        "page[size]": str(limit),
        "page[number]": "1",
    }
    return f"{API_URL}?{urllib.parse.urlencode(params)}"


def fetch(url: str) -> dict:
    req = urllib.request.Request(url, headers=HEADERS)
    with urllib.request.urlopen(req, timeout=20) as resp:
        return json.loads(resp.read().decode("utf-8"))


def index_included(payload: dict):
    """Build lookup tables for included entity/language resources by (type, id)."""
    lookup = {}
    for item in payload.get("included", []):
        lookup[(item.get("type"), item.get("id"))] = item.get("attributes", {})
    return lookup


def get_related(rel, lookup, rel_type):
    """Resolve a to-one relationship to its attributes dict, or {}."""
    data = rel.get(rel_type, {}).get("data")
    if not data:
        return {}
    return lookup.get((data.get("type"), data.get("id")), {})


def rows_from_payload(payload: dict, base_url: str = "https://filings.xbrl.org"):
    lookup = index_included(payload)
    rows = []
    for item in payload.get("data", []):
        attrs = item.get("attributes", {})
        rels = item.get("relationships", {})

        entity_attrs = get_related(rels, lookup, "entity")
        language_attrs = get_related(rels, lookup, "language")

        entity_name = entity_attrs.get("name", "—")
        country = attrs.get("country", entity_attrs.get("country", "—")) or "—"
        language = language_attrs.get("name", attrs.get("language", "—")) or "—"
        period_end = attrs.get("period_end", "—") or "—"
        date_added = attrs.get("date_added", "—") or "—"

        # "View" link — the filings.xbrl.org viewer URL for this filing
        filing_id = item.get("id", "")
        view_url = f"{base_url}/{filing_id}" if filing_id else "—"

        rows.append([entity_name, language, country, period_end, date_added, view_url])
    return rows


def print_table(rows, headers):
    if not rows:
        print("No filings found.")
        return

    widths = [len(h) for h in headers]
    for row in rows:
        for i, cell in enumerate(row):
            widths[i] = max(widths[i], len(str(cell)))

    def fmt_row(cells):
        return "  ".join(str(c).ljust(widths[i]) for i, c in enumerate(cells))

    print(fmt_row(headers))
    print("  ".join("-" * w for w in widths))
    for row in rows:
        print(fmt_row(row))


def main():
    parser = argparse.ArgumentParser(
        description="Look up ESEF/XBRL filings by ticker or entity name."
    )
    parser.add_argument(
        "query", help="Ticker, LEI, or (partial) company name to search for"
    )
    parser.add_argument(
        "--limit",
        type=int,
        default=20,
        help="Max number of filings to return (default 20)",
    )
    args = parser.parse_args()

    url = build_url(args.query, args.limit)

    try:
        payload = fetch(url)
    except urllib.error.HTTPError as e:
        print(f"HTTP error {e.code}: {e.reason}", file=sys.stderr)
        sys.exit(1)
    except urllib.error.URLError as e:
        print(f"Network error: {e.reason}", file=sys.stderr)
        sys.exit(1)

    rows = rows_from_payload(payload)
    headers = ["Entity", "Language", "Country", "Period Ending", "Date Added", "View"]
    print_table(rows, headers)
    print(f"\n{len(rows)} filing(s) found for '{args.query}'.")


if __name__ == "__main__":
    main()

File: crates/core/src/interface/external/xbrl.rs
use reqwest::blocking::Client;
use reqwest::Url;
use serde::Deserialize;
use serde_json::Value;

use crate::Error;

const API_URL: &str = "https://filings.xbrl.org/api/filings";
const BASE_URL: &str = "https://filings.xbrl.org";

#[derive(Debug, Clone)]
pub struct XbrlFiling {
    pub entity_name: String,
    pub language: String,
    pub country: String,
    pub period_end: String,
    pub date_added: String,
    pub view_url: String,
}

#[derive(Debug, Deserialize)]
struct JsonApiResource {
    id: String,
    #[serde(rename = "type")]
    resource_type: String,
    attributes: Option<Value>,
    relationships: Option<Value>,
}

#[derive(Debug, Deserialize)]
struct JsonApiResponse {
    data: Vec<JsonApiResource>,
    #[serde(default)]
    included: Vec<JsonApiResource>,
}

pub fn query_filings(queries: &[&str], limit: usize) -> Result<Vec<XbrlFiling>, Error> {
    let filter = build_filter(queries);
    let url = build_url(&filter, limit)?;

    let client = Client::new();
    let resp = client
        .get(url)
        .header(
            "User-Agent",
            "Mozilla/5.0 (X11; Linux x86_64; rv:150.0) Gecko/20100101 Firefox/150.0",
        )
        .header("Accept", "application/json, text/javascript, */*; q=0.01")
        .header("X-Requested-With", "XMLHttpRequest")
        .header("Referer", "https://filings.xbrl.org/")
        .send()
        .map_err(|e| Error::Http(e.to_string()))?;

    let payload: JsonApiResponse = resp
        .json()
        .map_err(|e| Error::Http(format!("failed to parse response: {e}")))?;

    Ok(flatten_filings(payload))
}

fn build_filter(queries: &[&str]) -> Vec<Value> {
    let mut or_conditions = Vec::with_capacity(queries.len() * 4);
    for &q in queries {
        or_conditions.push(serde_json::json!({"name": "entity.name", "op": "ilike", "val": format!("%{q}%")}));
        or_conditions.push(serde_json::json!({"name": "entity.identifier", "op": "eq", "val": q}));
        or_conditions.push(serde_json::json!({"name": "language.name", "op": "ilike", "val": format!("{q}%")}));
        or_conditions.push(serde_json::json!({"name": "sha256", "op": "eq", "val": q}));
    }
    vec![
        serde_json::json!({"or": or_conditions}),
        serde_json::json!({"name": "input_filing.filing_source.program", "op": "ne", "val": "UAIFRS"}),
    ]
}

fn build_url(filter: &[Value], limit: usize) -> Result<Url, Error> {
    let filter_json =
        serde_json::to_string(filter).map_err(|e| Error::Http(format!("filter serialization: {e}")))?;

    Url::parse_with_params(
        API_URL,
        &[
            ("include", "entity,language"),
            ("filter", &filter_json),
            ("sort", "-date_added"),
            ("page[size]", &limit.to_string()),
            ("page[number]", "1"),
        ],
    )
    .map_err(|e| Error::Http(format!("invalid URL: {e}")))
}

fn flatten_filings(payload: JsonApiResponse) -> Vec<XbrlFiling> {
    let lookup: Vec<(&str, &JsonApiResource)> = payload
        .included
        .iter()
        .map(|r| (r.resource_type.as_str(), r))
        .collect();

    payload
        .data
        .into_iter()
        .map(|item| {
            let attrs = item.attributes.as_ref();
            let rels = item.relationships.as_ref();

            let entity_attrs = resolve_related(rels, "entity", &lookup);
            let language_attrs = resolve_related(rels, "language", &lookup);

            let entity_name = entity_attrs
                .and_then(|a| a.get("name"))
                .and_then(|v| v.as_str())
                .unwrap_or("\u{2014}");
            let country = attrs
                .and_then(|a| a.get("country"))
                .and_then(|v| v.as_str())
                .or_else(|| entity_attrs.and_then(|a| a.get("country")).and_then(|v| v.as_str()))
                .unwrap_or("\u{2014}");
            let language = language_attrs
                .and_then(|a| a.get("name"))
                .and_then(|v| v.as_str())
                .or_else(|| attrs.and_then(|a| a.get("language")).and_then(|v| v.as_str()))
                .unwrap_or("\u{2014}");
            let period_end = attrs
                .and_then(|a| a.get("period_end"))
                .and_then(|v| v.as_str())
                .unwrap_or("\u{2014}");
            let date_added = attrs
                .and_then(|a| a.get("date_added"))
                .and_then(|v| v.as_str())
                .unwrap_or("\u{2014}");

            XbrlFiling {
                entity_name: entity_name.to_string(),
                language: language.to_string(),
                country: country.to_string(),
                period_end: period_end.to_string(),
                date_added: date_added.to_string(),
                view_url: format!("{BASE_URL}/{}", item.id),
            }
        })
        .collect()
}

fn resolve_related<'a>(
    rels: Option<&Value>,
    name: &str,
    lookup: &[(&str, &'a JsonApiResource)],
) -> Option<&'a Value> {
    let data = rels?
        .get(name)?
        .get("data")?;
    let rel_type = data.get("type")?.as_str()?;
    let rel_id = data.get("id")?.as_str()?;
    lookup
        .iter()
        .find(|(t, r)| *t == rel_type && r.id == rel_id)
        .and_then(|(_, r)| r.attributes.as_ref())
}
